posts/dumb_things_with_postgres.md

```diff
@@ -0,0 +1,77 @@
+# Doing dumb things with Postgres
+
+A couple of months ago I gave a talk about PostgreSQL, and specifically using the Array datatype at [London Ruby User Group](https://skillsmatter.com/skillscasts/9865-london-ruby-lightning-talks) and I wanted to take some time to dive into this in more detail.
+
+I've been fascinated with the Array datatype from Postgres and using it in Rails since it became usable in 2014, at the time I worked for a company that used Postgres for all of the new projects and several came up where it fitted in nicely with the work we were doing. Based on this I decided to give it a go and see what benefits we could gain and also try to work out any major pitfalls.
+
+The array datatype works exactly as expected, instead of storing an string or integer it instead allows you to save arrays of these types. We used these for all sorts of things; from storing article tags and CSS classes to storing multiple associations without the use of a joining table. The latter usage is the area I became most involved an interested in. Given the product we worked on would be mainly used in a read heavy manner the additional overhead of joining tables could occassionally cause bottlenecks, especially on more complicated projects, by leveraging arrays, Rails caching and some code that I've come to regret we were able to mitigate these.
+
+Example:
+
+You have a series of news articles that can have associated tags that we use to find other articles, normally you would do this by implementing this;
+
+Articles Table -> Taggings joining table -> Tags Table
+
+# PUT DIAGRAM/screenshot ABOVE
+
+Using the array datatype you could instead have;
+
+Articles Table -> Tags Table
+
+# PUT DIAGRAM/screenshot ABOVE
+
+Doing this provides us a couple of advantages, with caching this can be noticeably faster than joining the 3 seperate tables and provides us a clean datastructure that we don't have to manipulate or `pluck` from to get the thing we care about, in this case the tag IDs. It also allows us to easily execute some very basic SQL to find articles that have the same tags as the current one being shown to the site user.
+
+```ruby
+class Article < ActiveRecord::Base
+  # all the other logic
+
+  def associated_articles
+    Article.where(id != self.id).where("tags = ANY(#{self.tags})")
```

Comment:
Good on you, never run the stuff I write. always scared of blogging broken code.

Yeah `where.not(id: self.id)` I guess.